### DISCUSSION TITLE
How to integrate an API which needs several requests to achieve one task?

### DISCUSSION THREAD

**Discussion by kiryph** [saloonphp/saloon#54] [Q&A]
I am wondering what a canonical way would be to integrate an API where I have to make several requests to get the desired result.

My specific case is the API of [beds24.com](https://beds24.com/), a booking & property management software for hotels, B&Bs, holiday homes, ...

The API is documented under [api.beds24.com](https://api.beds24.com/)

When I like to fetch all arrivals for a specific date, I have to make a request for each property, e.g. one request for holiday apartment A and a second request for holiday apartment B.

When all requests are made, I want a single array of all arrivals.

The API endpoint is

https://api.beds24.com/json/getBookings

The request contains a json body as noted:
```json
{
    "authentication": {
        "apiKey": "apiKeyAsSetInAccountSettings",
        "propKey": "propKeyAsSetForTheProperty"
    },
    "includeInvoice": false,
    "includeInfoItems": false
}
```

I use Laravel 9. The API keys are in my `.env` and I get them via the config file where `propKeys` contains an array of `propIds` as keys (here `123456` and `789012`) and `propKey` as values:

<details><summary><tt>config/beds24.php</tt></summary>

```php
<?php

return [
    'apiKey' => env('BEDS24_APIKEY'),
    'propKeys' => [

        '123456' => env('BEDS24_PROPKEY_123456'),
        '789012' => env('BEDS24_PROPKEY_789012'),
    ],
];
```
</details>

So far I have written a Saloon connector for beds24:

<details> <summary><tt>app/Http/Integrations/beds24/Beds24Connector.php</tt></summary>

```php
<?php

namespace App\Http\Integrations\beds24;

use Sammyjo20\Saloon\Http\SaloonConnector;
use Sammyjo20\Saloon\Traits\Plugins\AcceptsJson;
use Sammyjo20\Saloon\Traits\Plugins\HasJsonBody;

use App\Http\Integrations\beds24\Requests\GetProperties;
use App\Http\Integrations\beds24\Requests\GetProperty;


class Beds24Connector extends SaloonConnector
{
    use AcceptsJson;
    use hasJsonBody;

    protected array $requests = [
        'get_properties' => GetProperties::class,
        'get_property' => GetProperty::class,
    ];

    public function defineBaseUrl(): string
    {
        return 'https://api.beds24.com/json';
    }

    public function defaultData(): array
    {
        return [
            'authentication' => [
                'apiKey' => config("beds24.apiKey"),
            ]
        ];
    }
}
```
</details>
and two requests  

1. get a list of all properties

<details> <summary><tt>app/Http/Integrations/beds24/requests/GetProperties.php</tt></summary>

```php
<?php

namespace App\Http\Integrations\beds24\Requests;

use App\Http\Integrations\Beds24\Beds24Connector;

use Sammyjo20\Saloon\Constants\Saloon;
use Sammyjo20\Saloon\Http\SaloonRequest;


class GetProperties extends SaloonRequest
{

    protected ?string $connector = Beds24Connector::class;

    protected ?string $method = Saloon::POST;

    public function defineEndpoint(): string
    {
        return '/getProperties';
    }

}
```
</details>

2. get details to a single property specified by the `propId`

<details> <summary><tt>app/Http/Integrations/beds24/requests/GetProperty.php</tt></summary>

```php
<?php

namespace App\Http\Integrations\beds24\Requests;

use App\Http\Integrations\Beds24\Beds24Connector;

use Sammyjo20\Saloon\Constants\Saloon;
use Sammyjo20\Saloon\Http\SaloonRequest;
use Sammyjo20\Saloon\Traits\Plugins\HasJsonBody;


class GetProperty extends SaloonRequest
{

    use HasJsonBody;

    protected ?string $connector = Beds24Connector::class;

    protected ?string $method = Saloon::POST;

    public function defineEndpoint(): string
    {
        return '/getProperty';
    }

    public function defaultData(): array
    {
        return array_merge_recursive(
                $this->getConnector()->defaultData(),
                [
                    'authentication' => [ 
                        'propKey' => config('beds24.propKeys')[$this->propId],
                    ],
                    'includeRooms' => false,
                    'includeRoomUnits' => false,
                    'includeAccountAccess' => false 
                ]
            );
    }

    public function __construct(
        public string $propId
    ){}

}
```
</details>

For test purposes, I added api routes:

<details><summary><tt>routes/api.php</tt></summary>

```php
Route::get('getProperties', function() {

    $request = Beds24Connector::get_properties();

    $response = $request->send();

    return $response;
});

Route::get('getProperties/{propId}', function($propId) {

    if (array_key_exists($propId, config('beds24.propKeys'))) {
        $request = Beds24Connector::get_property($propId);
        $response = $request->send();
        return $response;
    } else {
        abort( response()->json('no propKey for propID found', 404) );
    }
});
```
</details>

So now I want to fetch bookings. My question is, how can I write a meta request like `get_bookings(checkin_date, checkout_date)` to fetch all bookings within this period for all `propIds`.

Naively, I would write:

```php
Route::get('getBookings', function() {
  $bookings = [];
  foreach ( config('beds24.propKeys') as $propId => $propKey) {
          $request = Beds24Connector::get_bookings($propId);
          $response = $request->send();
          array_merge($bookings, $response->json());
  }
  return $bookings;
});
```

However, beds24 has a default limit of 60 API calls per 5 minutes for an account which I want to make sure I strictly follow. If I monitor API calls correctly, I can delay the next request, so I do not run into problems with beds24.

They say

> If you put a delay of 5 seconds between calls you should have no problems.

I ignore this hint because I have to query usually 5-10 different properties and adding (n-1)*5 sec to the method makes it too slow. As long as I stick to the limit of 60 API calls per 5 minutes, beds24 does not complain.
 
Before I tried Saloon, I used guzzle directly and created a pool of requests (https://docs.guzzlephp.org/en/stable/quickstart.html#concurrent-requests).

**How can I create a pool of requests with Saloon and run them concurrently?**

Since there is a discussion of including a rate limiting plugin, I was wondering how multiple, similarly structured requests fit in and the planned rate limiting plugin can ensure the limit of 60 API calls per 5 minutes not only per single request but also for a set of requests.

**UPDATE: Request Pool with Guzzle**
using the design pattern ‘repository’

```php
<?php

namespace App\Repositories;

use GuzzleHttp\Client;
use GuzzleHttp\RequestOptions;
use GuzzleHttp\Pool;
use GuzzleHttp\Psr7\Response;

use App\Contracts\BookingRepository;


class Beds24BookingRepository implements BookingRepository
{
    public function attendees($date)
    {
        $auth = array();
        $auth['apiKey'] = config('services.beds24.apiKey');

        $data = array();

        /* Reduce transferred data */
        $data['arrivalFrom'] = "20000101"; // necessary to overwrite default which points to today
        $data['arrivalTo'] = $date->format('Ymd');
        $data['departureFrom'] = $date->format('Ymd');
        $data['includeInvoice'] = false;
        $data['includeInfoItems'] = true;
        $data['includeInfoItemsConverted'] = true;
        $data['status'] = '1';

        $client = new Client();
        $requests = function () use ($client, $auth, $data) {

            foreach (config('services.beds24.propKeys') as $value) {

                $auth['propKey'] = $value;
                $data['authentication'] = $auth;
                $headers = [RequestOptions::JSON => $data ];
                yield function() use ($client, $headers) {
                    $url = config('services.beds24.apiUrl').'getBookings';
                    return $client->postAsync($url, $headers);
                };
            }
        };

        $result = array();
        $pool = new Pool($client, $requests(), [
            'concurrency' => count(config('services.beds24.propKeys')),
            'fulfilled' => function (Response $response) use (&$result) {
                if ($response->getStatusCode() == 200) {
                    $res = json_decode($response->getBody(), true);
                    $result = array_merge($result, $res);
                  }
            },
        ]);

        $promise = $pool->promise();
        $promise->wait();

        if (array_key_exists('errorCode', $result)) {
            return $result['error'];
        }

        return $result;
    }
}
```

**Comment by Sammyjo20** [comment]
Hey @kiryph thank you for your post and all the detail you included. It's very useful.

I see what you are trying to do, currently Saloon does not support request pooling or asynchronous requests. This is something I would absolutely be interested in working on as Async requests is probably the only part of Guzzle that Saloon doesn't touch right now! 

I could probably write something custom using Guzzle's existing promise architecture. Would you expect to receive an array of responses in the order of requests that were pooled?

**Reply by kiryph** [reply]
> Would you expect to receive an array of responses in the order of requests that were pooled?

I have not thought about whether maintaining order is always necessary. In the mentioned use case the order is not important.

I remember using [GNU Parallel](https://www.gnu.org/software/parallel/) where the results should be in the same order as the input because the output did not contain anymore the input value.

Async networking in general does not maintain order and you would need to order them after all requests are complete or fill an array non-sequentially. Maybe one can add this functionality when someone needs it.

**Reply by juse-less** [reply]
> I could probably write something custom using Guzzle's existing promise architecture. Would you expect to receive an array of responses in the order of requests that were pooled?

I, just now, quickly threw this together as I needed it myself, and could send a PR.
I didn't implement the pooling capability. Just the Async method, which was very quick with Saloon.
It's just that I had to override a few files in the project (RequestManager, Request, and Connector), to accompany the regular `send()` method.
Given I had to replace some files (mostly RequestManager), it would be a lot better if it can be merged into the library.

**Reply by Sammyjo20** [reply]
Thanks for contributing, I will take a look at your Fork, yes I actually played around with Async too!

**Reply by juse-less** [reply]
It just occurred to me that I don't set the synchronous option, like Guzzle does internally in their `send()` method.
Not sure what the implications might be. It seems it's only used in the Proxy that wraps a 'synchronous' request.

FWIW, though, I haven't had any issues doing the same with Laravels HTTP Client for the past half year.

**Reply by Sammyjo20** [reply]
So if Saloon had async requests, that would sort the issue? When I did some testing myself it's quite easy to get async requests working but pools are a bigger job.

**Reply by juse-less** [reply]
I feel like I hijacked this one a bit.
I thought I'd mention I've implemented at least a PoC for async requests, since I saw you mentioning it.

For me, personally, having async requests in the library would definitely solve a lot for me, and I wouldn't have to override classes.

For this particular use case, it sounds like @kiryph would optimally need something more to property handle rate limiting and pooling.

FWIW, though, since I saw he's using Laravel..
What I've done to rate limit the app I'm developing, is to create a rate limiter in Laravel to [rate limit Jobs](https://laravel.com/docs/9.x/queues#rate-limiting).
Then do N amount of requests per Job, which effectively keeps it in check with external API's requirements.
This was a quick and dirty way to wait in-between the requests.

**Reply by Sammyjo20** [reply]
Hey all,

I have created a draft PR for my best implementation of Async requests, but I need your opinion on something.

Firstly, take a look here at the PR: https://github.com/Sammyjo20/Saloon/pull/56

Typically, Saloon will return a response with the error inside of the response, however with promises, if an error occurs - Saloon will return the response in the rejection chain instead. Should this be the case? Or should we only reserve rejection for exceptions like ConnectException and always return responses, even if errored in the resolve callback?

With this in place, it shouldn't be too hard to do Pools of requests.

**Reply by Sammyjo20** [reply]
@kiryph do you mind sharing your old Guzzle implementation of pools with this use case? I’m going to build the Saloon pools API pretty much the same way as Guzzle but if I can work from a concrete example that will help.

**Reply by kiryph** [reply]
@Sammyjo20 I have updated my question with a part of my previous usage of a Guzzle pool.

**Comment by Sammyjo20** [comment]
I have introduced support for asynchronous request in v1.1 https://github.com/Sammyjo20/Saloon/releases/tag/v1.1.0

I will take a look at making Saloon's own pooling since Guzzle's pools require a single client instance where Saloon's requests create a client per-request.

**Comment by Sammyjo20** [comment]
Hey, I am still looking into request pools. I have a branch that I am working on and I have a working concept, however it's not quite working as I expect, as it's not concurrent - but everything else seems to work. Once I crack that, hopefully I can bring it into Saloon (with a nice syntax too!)

Something like:

```php
$pool = Pool::make(array|callable $requests)
    ->onSuccess(fn () => ...)
    ->onFailure(fn () => ...)
    ->concurrency(10);
```

**Comment by Sammyjo20** [comment]
Right! With a fresh start to today, I did some more research. So it looks like Guzzle's asynchronous requests uses CURL's multi-handle,  which basically means it keeps a connection open and sends multiple requests with one connection open.

This is achieved because you would have one Guzzle client sending multiple asynchronous requests, so I assume Guzzle will keep the connection open until all requests have been processed. This is super cool! However Saloon works slightly different.

This is Saloon's flow:

1. You create a request class
2. You run the `sendAsync()` method
3. Every request builds a new Guzzle Client, and passes in the handlers/middleware 
4. Request is sent

This set up is great for flexibility because it allows you to define per-request middleware, which is something Guzzle does not offer out of the box. The disadvantage of this is the fact that a new Guzzle client is booted up for every request.

I ran a little test, the first screenshot is Guzzle's pooling:

<img width="952" alt="Screen Shot 2022-04-30 at 09 07 47" src="https://user-images.githubusercontent.com/29132017/166097507-18ec372f-e286-4d23-aa41-34a883917ac7.png">

As you can see, the connection is closed after all the requests have been fulfilled. Here's a screenshot of Saloon's pooling:

<img width="952" alt="Screen Shot 2022-04-30 at 09 08 33" src="https://user-images.githubusercontent.com/29132017/166097532-6a4fe032-44a9-4d8b-b965-e81fc19acde4.png">

As you can see, the connection is closed after every request, so it's not asynchronous at all! 😬  You get the benefit of the promise, but that's about it.

So I'm thinking we could still release Pools, however it would have to work different. For Saloon pools, you will only be able to define handlers on the connector and not the request. This means request handlers will be completely disabled. You could still use caching, but it would mean you have to apply it to the connector. The pool would then have to accept a connector, and boot up a Guzzle client once and send multiple asynchronous requests. 

This does reduce Saloon's functionality, but would allow you to use real asynchronous requests.

Let's discuss this, what do people think? If Guzzle can support a different handler stack per request that would be awesome!

**Reply by juse-less** [reply]
First off - thanks for providing the findings of your research and testing. This is very valuable, at least for me, Saloon or not.

Something I thought of right away, regarding Guzzle not supporting per-request handlers/middlewares is that.. maybe you could queue up registered 'global' middlewares, but also add a `Middleware::mapRequest()`, as well as `Middleware::mapResponse()`, which would check every request (and request/response pairs, respectively) for defined middlewares, and then apply them in these sort of wrappers from those `Middleware::map*()` handlers.
If the request/response doesn't have anything defined, you can just pass it on to the next middleware in the stack.
```php
if (/* $request/$response has middleware ... */) {
    // pipe the $request/$response through them
}

return $next(/* $request/$response */);
```
Then I think you'd have support for both global, and per-request middlewares, and support that CurlMultiHandler.

https://docs.guzzlephp.org/en/stable/handlers-and-middleware.html#handlerstack

I don't have a need for different PSR-7 libraries, and pretty much exclusively use Guzzle myself, but would still like to link to this discussion, so that it's not forgotten and sort of collides with this Guzzle-feature.
https://github.com/Sammyjo20/Saloon/discussions/50

**Reply by Sammyjo20** [reply]
Ah! I think that's what I've been looking for. I will have a play around with that and see if I can implement it into Saloon. Thanks for this, very useful.

It would still be built inside of the RequestManager, so I would say it's not impossible to achieve #50 in the future, but maybe for a v2 or v3 thing, I was also thinking of using something like HTTPlug.

**Reply by juse-less** [reply]
Let me know how it goes. 🙂
With this pooling, it would be a lot easier to also handle paginated resources.
Like.. having a pagination middleware, that just takes the values of the request and seamlessly just increases the page number.

I've previously been looking into this repo for something like that, but haven't had the time to make something of it.
Maybe it could give you ideas, as well.
https://github.com/cerbero90/lazy-json-pages

**Comment by Sammyjo20** [comment]
Great news, I have a working prototype of a new RequestManager that will be able to send multiple requests asynchronously AND the ability to have per-request middleware, which is absolutely massive. It means you can have middleware on the connector, then a different set of handlers on each request. One request could have a cache handler, the other one ins't cached at all.

I achieved it by creating an X-Saloon-ID header on each request, and the for each handler wrapping it in a closure that checks if the current request matches the ID on the header. If it does, it will run the handler, otherwise it will continue. 

At the end of the handler stack, it removes the X-Saloon-ID header, just so it isn't actually sent to the internet.

To be honest, the changes I'm making will be breaking to the underlying SaloonRequest class, and I want to overhaul the request manager anyway, so I will probably release this as part of Saloon v2, even though v1 hasn't been out for very long.

I think the waiting time would be between 3-5 weeks to have this sorted, but I'm very excited to be able to introduce functionality that Guzzle doesn't have under the hood!

I'm also very excited to introduce concurrent requests as well.

**Reply by kiryph** [reply]
These are exciting news. I am really looking forward. Thank you very much for adding this and for creating Saloon in the first place.

**Comment by Sammyjo20** [comment]
Hey there 👋🏼 

Just wanted to let y'all know that I can happily confirm that **real** asynchronous support will be added in Saloon v2. It will use a similar "pooling" class, but I have rebuilt it in a way where we can keep a request open throughout all requests.

I have also created Saloon's own Middleware pipeline, so you can have a custom middleware pipeline per-request/connector and it will still be executed properly which is something that we couldn't achieve before with Guzzle.

**Reply by juse-less** [reply]
I don't mean to stress. But.. do you possibly have an ETA on when you'd think v2 may be released? 🙂
Can't wait for v2 of Saloon (I saw the branch the other day), and Spatie's Laravel Data, is released. 👀

**Reply by Sammyjo20** [reply]
It's going really well so far. Knowing me, I get excited and have a couple of really good coding sessions. It could be as early as 3 weeks from now, but it could be anywhere between 3 and 6 weeks. I will keep you updated! Appreciate the support!

**Reply by michabbb** [reply]
hi @Sammyjo20, I would like to wait for v2 before I start using this package, do you have any idea when you will release v2? 😊

**Reply by Sammyjo20** [reply]
Hey @michabbb it'll be very early next year but the first beta will be out in a couple of weeks! Watch the repo for releases :)

**Reply by Sammyjo20** [reply]
Hey @michabbb just wanted to let you know v2 is out and it comes with full asynchronous request support alongside request pooling and concurrency 👍

**Reply by michabbb** [reply]
Thanks, I have subscribed to this repo, and I have seen your hard work here the last weeks, great job 😍👍

**Reply by Sammyjo20** [reply]
Brilliant, enjoy!!
